Added delete of value labels to the Prepare > Column: Factor > View Levels/Labels dialogue using delete key in column metadata

N-thony opened the pull request:
Fixes (partially) #7073 
This is still a work in progress.
@dannyparsons I need your help here. Could we have a call? Thanks

N-thony commented:
> Fixes (partially) #7073 This is still a work in progress. @dannyparsons I need your help here. Could we have a call? Thanks

@dannyparsons could we have a discussion on this?

N-thony commented:
> @N-thony that seems fine, as far as it goes, until you have discussed with @dannyparsons Keep asking him! a) I notice that Ok isn't enabled for the data frame, when you first use the delete option. b) If you go to the columns, and choose a variable, and return to the data frame, then it is enabled. But it currently delivers the same command for the columns, not anything for the whole data frame. But that will change anyway. c) Related, if easy, will be to delete a label from the metadata. I notice that this works in principle, but not in practice. It delivers the following command:
> 
> ```
> # Edited variables metadata value
> data_book$append_to_variables_metadata(data_name ="survey",col_names = "fert",property="labels",new_val="")
> ```
> 
> Isn't that the sort of thing you want as your function. I also tried with:
> 
> `data_book$append_to_variables_metadata(data_name ="survey",col_names = "fert",property="labels",new_val="NA") ` Maybe you could be trying that as the code in your dialogue. By the way it didn't seem to work, through the metadata window. Perhaps it will work for you, or that's what you could investigate while waiting for Danny?
> 
> Once sorted, then we should also see what needs to change in the metadata window. Currently it just seems to mess up the factor!
@rdstern thanks for review though this is pending and will be ready for review after the help from Danny. 
@dannyparsons could we discuss this or maybe guide me probably on the backend side? Thanks.

N-thony commented:
@rdstern @shadrackkibet I have added the delete labels feature on View Levels/Labels dialogue, and in addition I have added the possibility of deleting the labels from `Delete` keyboard. Please take a look. Thanks.

rdstern reviewed (changes requested):
@N-thony I don't know if you  checked the results of what you have done with the delete?
I first tried deleting all in the survey data.  Then I found that the levels/Labels dialogue doesn't work at all.  And the Labels has = NA for all variables, while perhaps it should be NA.
When I started again and deleted just the fert (which I had made a labelled factor, Then I was still ok examining the other factors with the Levels/Labels, but not the fert that I had deleted.
When choosing the variables I wonder if the data selector could just have the variables that have value labels to delete?  (Perhaps ideally where there are none, there might be a message saying No Variables have Value Labels to Delete.

rdstern commented:
I also notice that when there are no labelled variables, then that item is absent in the metadata.  For example, look at the survey data when imported.  There is no Value Label entry.  If you then make fert into a factor, then it has value labels and the entry is provided.  I suggest if you delete value labels from the data frame, then it might delete the entry completely, rather than keeping it with NA in every variable.

N-thony commented:
> @N-thony I don't know if you checked the results of what you have done with the delete? I first tried deleting all in the survey data. Then I found that the levels/Labels dialogue doesn't work at all. And the Labels has = NA for all variables, while perhaps it should be NA. When I started again and deleted just the fert (which I had made a labelled factor, Then I was still ok examining the other factors with the Levels/Labels, but not the fert that I had deleted. When choosing the variables I wonder if the data selector could just have the variables that have value labels to delete? (Perhaps ideally where there are none, there might be a message saying No Variables have Value Labels to Delete.

@rdstern of course, I did the test and it is very trivial to notice that, I was wondering why it was adding `=` even if you try to delete the label manually on the grid by replacing with nothing or  `NA` it will change it to `= NA`, I had the check on that with @shadrackkibet and I was thinking that the R function need to be improved

![image](https://user-images.githubusercontent.com/68591383/157166609-c58c6bb9-4567-4bb7-8c12-ffbfeef62a31.png)

![image](https://user-images.githubusercontent.com/68591383/157166563-18464607-82f3-4c32-8e1d-6f22c6f36a4a.png)

N-thony commented:
@rdstern @shadrackkibet I have resolved the comment above and made a small change in the used R function here.
@rdstern could you re-test? Thanks.
@shadrackkibet could you also test and check in the R function? Thanks

rdstern reviewed (dismissed):
I am approving, but add some items I noticed.
a) Ideally when you choose Selected Variables, the data selector would say Labelled Variables and just give a list of them - rather than all variables.  I don't know if that is easy.  I am hoping that @shadrackkibet Select feature could also select all labelled variables in a data frame?
b) I don't know if the editing of the metadata is in the same pull request.  There I noticed that the Delete button works nicely.  I wouldn't mind if the right-click, when you choose one-or-more cells here also had Delete.  Currently it just has Help.
c) One item (I presume) to turn off, is that you can currently edit within those fields.  There you can really mess up the labelling.  That needs to be turned off.  So, then there is nothing you can do to value labels in that grid, excpet delete them.

lloyddewit commented on instat/dlgViewLabelsAndLevels.vb:
```diff
@@ -132,7 +180,38 @@ Public Class dlgViewFactorLabels
-    Private Sub ucrReceiverFactorColumns_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverVariables.ControlContentsChanged, ucrChkShowFrequencies.ControlContentsChanged, ucrChkShowLabels.ControlContentsChanged, ucrChkShowPercentage.ControlContentsChanged, ucrChkShowType.ControlContentsChanged, ucrChkShowValues.ControlContentsChanged
+    Private Sub SetReceiverVariableVisible()
+        If rdoDeleteValueLabels.Checked Then
+            ucrReceiverVariables.Visible = If(rdoSelectedColumn.Checked, True, False)
```
simplify?
```suggestion
            ucrReceiverVariables.Visible = rdoSelectedColumn.Checked
```

lloyddewit commented on instat/ucrColumnMetadata.vb:
```diff
@@ -122,6 +123,26 @@ Public Class ucrColumnMetadata
+                frmMain.clsRLink.RunScript(clsDeleteLabelsFunction.ToScript())
+            End If
+        Else
+            MsgBox("Deleting cells is currently disabled. This feature will be included in future versions." & Environment.NewLine & "To remove a cell's value, replace the value with NA.", MsgBoxStyle.Information, "Cannot delete cells.")
```
add line breaks to make more readable

lloyddewit reviewed (changes requested):
@N-thony thanks, just a few questions

N-thony commented:
> Event DeleteLabels(strColumnName As String)

@rdstern just a typo when pushing, could you test again?

N-thony commented on instat/Interface/IColumnMetaDataGrid.vb:
```diff
@@ -19,6 +19,8 @@ Public Interface IColumnMetaDataGrid
 
     Event EditValue(iRow As Integer, strColumnName As String, strPreviousValue As String, newValue As Object)
 
+    Event DeleteLabels(strColumnName As String)
```
This is more specific for the metadata grid, the reason why that is specific.

lloyddewit replied:
thanks, makes sense

N-thony commented:
@lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks

rdstern reviewed (approved):
I am still happy and relieved to approve this.  If possible I would still like the selector to indicate and just show the Labelled Variables.  But that may come later

lloyddewit commented on instat/dlgLabelsLevels.vb:
```diff
@@ -114,21 +117,30 @@ Public Class dlgLabelsLevels
-        Dim iMissingValue As Integer
-        Dim clsGetColumnFunction As RFunction = ucrReceiverLabels.GetVariables()
+    Private Sub CountLevels()
+        clsGetColumnFunction = ucrReceiverLabels.GetVariables()
```
@N-thony this still needs resolving

lloyddewit commented on instat/dlgLabelsLevels.vb:
```diff
@@ -19,6 +19,8 @@ Public Class dlgLabelsLevels
     Private bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private clsViewLabelsFunction, clsSumCountMissingFunction As New RFunction
+    Private iMissingValue As Integer
```
@N-thony this still needs resolving

lloyddewit commented on instat/dlgLabelsLevels.vb:
```diff
@@ -19,6 +19,8 @@ Public Class dlgLabelsLevels
     Private bReset As Boolean = True
     Private clsViewLabelsFunction, clsSumCountMissingFunction As New RFunction
+    Private iMissingValue As Integer
+    Private clsGetColumnFunction As RFunction
```
@N-thony this still needs resolving

lloyddewit commented:
> @lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks

@N-thony there are many comments above still unresolved. I started going through them and stating '@N-thony this still needs resolving' buth there are too many. Please go through all the comments again (some may be hidden) and ensure that you've pushed all your changes, thanks

N-thony commented:
> > @lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks
> 
> @N-thony there are many comments above still unresolved. I started going through them and stating '@N-thony this still needs resolving' buth there are too many. Please go through all the comments again (some may be hidden) and ensure that you've pushed all your changes, thanks

I think all is resolved now.

lloyddewit reviewed (approved):
@N-thony all my comments are now resolved, thanks

lloyddewit commented:
@rdstern Please could you test again?
@shadrackkibet Would you also like to test/review this PR or may we go ahead and merge?
Thanks

lloyddewit commented:
> I am still happy and relieved to approve this. If possible I would still like the selector to indicate and just show the Labelled Variables. But that may come later

@N-thony Will you implement this in a separate PR?

shadrackkibet commented:
@lloyddewit we just had discussions about this. Just a few things to sort out then we can merge.

shadrackkibet reviewed (approved):
@lloyddewit over to you.

lloyddewit commented:
@rdstern If you're able to retest and approve then we can merge, thanks

rdstern reviewed (approved):
Great
